docs/how-to/tune-settings.md
# How to tune etcd settings

Charmed etcd includes default configuration settings that work well in most deployment settings. In some cases though,
it might be required to adjust the configuration. This guide will show how to tune charmed etcd to match specific requirements.

## Network latency

In networks with high latency or across regions, the default configuration can trigger unnecessary leader elections and disrupt the cluster. 
On the other hand, in deployments where fast failover matters, a longer election timeout can delay recovery when a leader fails.

Charmed etcd lets you adjust the settings for `heartbeat-interval` and `election-timeout`, in order to:
- Tolerate slower networks without constant leader re-elections
- Reduce overhead in resource-constrained environments by lowering heartbeat frequency
- Speed up failover when low latency is critical

In the following example, we will set `heartbeat-interval` to 300 milliseconds and `election-timeout` to 3 seconds.

You can either configure these settings at deploy time by adding the configuration to your deploy-command, for example:

```text
juju deploy charmed-etcd -n 3 --config heartbeat-interval=300 --config election-timeout=3000
```

Or you can set them on your existing charmed etcd application by running `juju config <application-name>...`, for example:

```text
juju config charmed-etcd heartbeat-interval=300 election-timeout=3000 
```

Charmed etcd will validate the configuration values and apply them to all units in the deployment (with a rolling restart
in case of an existing deployment). 

For more information about the allowed values, you can inspect the description of these settings by running `juju config charmed-etcd:`

```text
[...]
charm: charmed-etcd
settings: 
  election-timeout: 
    default: 1000
    description: How long, in milliseconds, a follower node will go without hearing
      a heartbeat before attempting to become leader itself. The value should be at
      least 10x the value of <heartbeat-interval>, and must not exceed 50000.
    source: user
    type: int
    value: 3000
  heartbeat-interval: 
    default: 100
    description: The frequency, in milliseconds, with which the leader will notify
      followers that it is still the leader. For best practices, the parameter should
      be set around round-trip time between members. The value must be between 10
      and 5000 at most.
    source: user
    type: int
    value: 300
[...]
```

If the configuration values are invalid, charmed etcd will display a status message in `juju status`:

```text
Model  Controller      Cloud/Region         Version  SLA          Timestamp
etcd   dev-controller  localhost/localhost  3.6.5    unsupported  11:23:13Z

App           Version  Status   Scale  Charm         Channel   Rev  Exposed  Message
charmed-etcd           blocked      3  charmed-etcd  3.6/edge   66  no       Invalid values set for the config options: 'election-timeout', 'heartbeat-interval'

Unit             Workload  Agent  Machine  Public address  Ports     Message
charmed-etcd/0*  blocked   idle   9        10.136.16.209   2379/tcp  Invalid values set for the config options: 'election-timeout', 'heartbeat-interval'
charmed-etcd/1   blocked   idle   10       10.136.16.121   2379/tcp  Invalid values set for the config options: 'election-timeout', 'heartbeat-interval'
charmed-etcd/2   blocked   idle   11       10.136.16.41    2379/tcp  Invalid values set for the config options: 'election-timeout', 'heartbeat-interval'

Machine  State    Address        Inst id         Base          AZ  Message
9        started  10.136.16.209  juju-335dd8-9   ubuntu@24.04      Running
10       started  10.136.16.121  juju-335dd8-10  ubuntu@24.04      Running
11       started  10.136.16.41   juju-335dd8-11  ubuntu@24.04      Running
```

Please refer to the [official etcd docs](https://etcd.io/docs/v3.6/tuning/#time-parameters) for more information on how
to decide the correct adjustments to the parameters for your specific requirements.

## Storage quota

By default, etcd imposes a storage quota of 2 GiB as the maximum permissible size of the etcd database
on each cluster member. While this limit may be sufficient for small-scale deployments, it might become
inadequate for bigger production clusters.

To adjust the storage quota of etcd, configure the `quota-backend-bytes` option. The following example
will set etcd's storage quota to 4 GiB:

```text
juju config charmed-etcd quota-backend-bytes="4000000000" 
```

The value of this configuration option should be sized according to the cluster's workload and available RAM.
When this configured limit is exceeded, the cluster will raise a `NOSPACE` alarm and enter a read-only mode,
rejecting all write requests.

Charmed etcd will validate the configuration value and apply it to all units in the deployment (with a rolling
restart in case of an existing deployment). The value needs to be `auto` or an integer representing the size 
in bytes.

If value is set to `auto`, the formula used is `min(100GiB, 0.9*memory, 0.9*data_storage_size)`.
A minimum of 100MiB is enforced.

For more information, please refer to the [official etcd docs](https://etcd.io/docs/v3.6/op-guide/maintenance/#space-quota).

## Auto-compaction

Charmed etcd makes use of auto-compaction to periodically compact the keyspace of the database
and avoid performance degradation and eventual storage space exhaustion. It applies default settings
for all deployments to compact the keyspace periodically and keep one hour of history.

To adjust these settings for specific use-cases, set the following configuration options:
- `auto-compaction-mode`
- `auto-compaction-retention`

With the `auto-compaction-mode` configuration, the compaction mode is defined. Possible values are 
"periodic" or "revision"; "periodic" for duration based retention, "revision" for revision number based retention.

The retention policy for the history compaction is configured with `auto-compaction-retention`. 
This setting specifies how much history to keep when compacting the keyspace. Supported time units
are "h" for hours or "m" for minutes in case of periodic compaction. If the compaction mode is set 
to "revision", this option defines how many revisions to keep.

As a safeguard to avoid storage exhaustion, a configuration value of `0` for `auto-compaction-retention`
will not be accepted.

## Certificate SANs configuration

```{attention}
Etcd's communication is strictly IP-based, both for peer-to-peer as well as client-to-server communication. 
Make sure to choose a TLS provider that supports IP SANs.
```

In X.509 TLS certificates, a `Subject Alternative Name` (SAN) allows a certificate subject to be associated with the 
service name and domain name components of a DNS record. If charmed etcd is deployed in an environment where the DNS 
resolution on the client side does not correspond to the DNS resolution on the server side, it might be required to add 
custom SANs to the TLS certificates used for client- and/or peer-communication.

See more about TLS in charmed etcd: [TLS encryption](tls/index.md)

### Add extra-sans

To add different IP addresses or hostnames to the SANs of charmed etcd's TLS certificates, configure the `certificate-extra-sans`
option. It is possible to add a comma-separated list of multiple values, as long as each of them is a valid IP address 
or hostname:

```text
juju config charmed-etcd certificate-extra-sans="10.241.9.34, etcd-production-cluster.mycompany.com"
```

If the configured sans are valid and not yet included in charmed etcd's certificates, it will automatically refresh
those.

```{caution}
Wildcards (`*`) are not allowed as part of the `certificate-extra-sans` configuration.
```

If it is required to include the unit number of each unit, this can be done by using the `{unit}` placeholder. For example,
a configuration of `certificate-extra-sans="etcd{unit}.my-external-domain.com"` would result in `etcd0.my-external-domain.com`
as an additional SAN in the TLS certificates for unit `charmed-etcd/0`.

### Add a domain

If a specific domain name is required for the TLS certificates used in etcd, configure the `client-certificate-domain`
and/or `peer-certificate-domain` option. The configuration value should match the allowed value from the TLS provider.

For example, if you want to expose your charmed etcd cluster externally and want to use client certificates for the 
domain `mycompany.com`, configure it like this:

```text
juju config charmed-etcd client-certificate-domain="mycompany.com"
```

This option can be configured for client and peer certificates separately. If the configured domain name is valid 
and not yet included in charmed etcd's certificates, it will automatically refresh those.


docs/how-to/backup-and-restore/configure-object-storage-provider.md
# How to configure an object storage provider
A backup for charmed etcd can be stored on any S3- or Azure-compatible storage. This includes:
- AWS S3
- Ceph with RadosGateway (see: [MicroCeph documentation](https://canonical-microceph.readthedocs-hosted.com/latest/snap/tutorial/get-started/#))
- Google Cloud Storage
- Azure Blob Storage

## Prerequisites
For configuration, you will need:
* the S3 or Azure account credentials
* the connection parameters

It is not required to set up a storage bucket or container in advance. This will be handled by the charmed operator: 
If it does not already exist, the charmed operator will create the storage bucket/container with the configured name
and the default permissions of your account.

```{caution}
Make sure to set up secure permissions on your storage account and bucket/container to avoid leaking sensitive information.
```

## Deploy integrator charm

The configuration and credentials management for the object storage are handled by an intermediate integrator charm. It 
is providing these information to charmed etcd. Depending on the type of object storage, you can use one of:
* `s3-integrator`
* `azure-storage-integrator`

### S3 storage
To deploy the `s3-integrator`, run:
```text
juju deploy s3-integrator
```

Once it is running, you should see an error message with `juju status`:
```text
Model  Controller      Cloud/Region         Version  SLA          Timestamp
etcd   dev-controller  localhost/localhost  3.6.5    unsupported  16:01:25Z

App            Version  Status   Scale  Charm          Channel   Rev  Exposed  Message
charmed-etcd            active       1  charmed-etcd   3.6/edge   49  no       
s3-integrator           blocked      1  s3-integrator  1/stable  145  no       Missing parameters: ['access-key', 'secret-key']

Unit              Workload  Agent  Machine  Public address  Ports  Message
charmed-etcd/0*   active    idle   2        10.198.26.225          
s3-integrator/0*  blocked   idle   1        10.198.26.200          Missing parameters: ['access-key', 'secret-key']
```

Now you need to configure the credentials for your S3 storage. Replace the placeholders and run the following command:
```text
juju run s3-integrator/leader sync-s3-credentials access-key=<access-key> secret-key=<secret-key> 
```

Finally, configure the connection parameters of your S3 storage to `s3-integrator`, like in the example shown below:
```text
juju config s3-integrator \
    endpoint="https://s3.us-west-2.amazonaws.com" \
    bucket="etcd" \
    path="/backups" \
    region="us-west-2"
```

For the full list of supported parameters, see [s3-integrator documentation](https://charmhub.io/s3-integrator/configurations).

### Azure storage
To deploy the `azure-storage-integrator`, run:
```text
juju deploy azure-storage-integrator --channel latest/edge
```

The `azure-storage-integrator` is currently only supported in the `latest/edge` channel.

Once it is running, configure the connection parameters of your Azure storage to `azure-storage-integrator`:
```text
juju config azure-storage-integrator storage-account=<Azure_storage_account> container=<Azure_storage_container>
```

The full list of parameters is available in the [documentation](https://charmhub.io/azure-storage-integrator/configurations).

To provide the credentials, create a secret, allow the `azure-storage-integrator` to read it and add it as a configuration:
```text
juju add-secret mysecret secret-key=<Azure_storage_key>
secret:d183rs296n7svmjrivp0

juju grant-secret mysecret azure-storage-integrator

juju config azure-storage-integrator credentials=secret:d183rs296n7svmjrivp0 
```

## Integrate with charmed-etcd
To provide the configuration and credentials of your object storage to charmed etcd, create a relation between the 
integrator charm and etcd:
```text
juju integrator <storage-integrator-charm> charmed-etcd
```

You can see a confirmation message in the log with `juju debug-log`:
```text
unit-charmed-etcd-0: 16:10:52 INFO unit.charmed-etcd/0.juju-log s3-credentials:3: Using existing bucket etcd
```

## Change storage configuration
If you want to change a connection parameter of your object storage, run the following command:
```text
juju config <storage-integrator-charm> <option>=<value>
```

src/managers/config.py
#!/usr/bin/env python3
# Copyright 2024 Canonical Ltd.
# See LICENSE file for licensing details.

"""Manager for handling configuration building + writing."""

import logging
import re
from pathlib import Path

import yaml
from data_platform_helpers.advanced_statuses.models import StatusObject
from data_platform_helpers.advanced_statuses.protocol import ManagerStatusProtocol
from data_platform_helpers.advanced_statuses.types import Scope
from ops.model import ConfigData

from core.cluster import ClusterState
from core.workload import WorkloadBase
from literals import (
    MAX_QUOTA_BACKEND_BYTES,
    METRICS_PORT,
    MIN_QUOTA_BACKEND_BYTES,
    RestoreStep,
    TLSState,
    TuningOptions,
)
from statuses import CharmStatuses, ConfigStatuses

logger = logging.getLogger(__name__)

WORKING_DIR = Path(__file__).absolute().parent


class ConfigManager(ManagerStatusProtocol):
    """Handle the configuration of etcd."""

    name: str = "config"
    state: ClusterState

    def __init__(
        self,
        state: ClusterState,
        workload: WorkloadBase,
        config: ConfigData,
    ):
        self.state = state
        self.workload = workload
        self.config = config
        self.config_file = workload.paths.config_file

    @property
    def config_properties(self) -> str:  # noqa: C901
        """Assemble the config properties.

        Returns:
            List of properties to be written to the config file.
        """
        with open(f"{WORKING_DIR}/config/etcd.conf.yml") as config:
            # load the config properties provided from the template in this repo
            # it does NOT load the template from disk in the charm unit
            # this is in order to avoid config drift
            config_properties = yaml.safe_load(config)

        config_properties["name"] = self.state.unit_server.member_name
        if (
            self.state.cluster.is_restore_in_progress
            and self.state.cluster.restore_instruction == RestoreStep.VERIFY
        ):
            # when verifying a backup restore, cluster configuration is only the local unit
            config_properties["initial-cluster-state"] = self.state.cluster.model.cluster_state
            config_properties["initial-cluster"] = self.state.unit_server.member_endpoint
        elif self.state.cluster.model.cluster_state:
            # regular situation: cluster is initialized and cluster configuration should be applied
            config_properties["initial-cluster-state"] = self.state.cluster.model.cluster_state
            config_properties["initial-cluster"] = self.state.cluster.model.cluster_members
        elif self.workload.exists(self.workload.paths.data_dir):
            # if no cluster state is available, but we find a database file
            # we force a new one-cluster-member with existing data
            # this is the case for storage reuse and `rebuild_cluster` workflows
            config_properties["force-new-cluster"] = True
        else:
            config_properties["initial-cluster-state"] = "new"
            # on very first cluster initialization, only the leader should be cluster member
            # all other units will be added to the cluster subsequently
            config_properties["initial-cluster"] = self.state.unit_server.member_endpoint
        config_properties["initial-advertise-peer-urls"] = self.state.unit_server.peer_url
        config_properties["listen-peer-urls"] = self.state.unit_server.peer_url
        config_properties["listen-client-urls"] = self.state.unit_server.client_url
        config_properties["advertise-client-urls"] = self.state.unit_server.client_url
        config_properties["listen-metrics-urls"] = (
            f"http://{self.state.unit_server.model.private_ip}:{METRICS_PORT}"
        )

        if self.are_tuning_parameters_valid():
            for option in TuningOptions:
                # take over the config value set by users
                value = self._get_tuning_config_value(option)
                config_properties[option.value] = value

        if self.state.unit_server.tls_client_state in [TLSState.TO_TLS, TLSState.TLS]:
            # replace http with https in listen-client-urls and advertise-client-urls
            config_properties["listen-client-urls"] = self.state.unit_server.client_url.replace(
                "http://", "https://"
            )
            config_properties["advertise-client-urls"] = self.state.unit_server.client_url.replace(
                "http://", "https://"
            )
            # set the client-transport-security
            config_properties["client-transport-security"] = {
                "cert-file": self.workload.paths.tls.client_cert.as_posix(),
                "key-file": self.workload.paths.tls.client_key.as_posix(),
                "client-cert-auth": True,
                "trusted-ca-file": self.workload.paths.tls.client_ca.as_posix(),
            }
        if self.state.unit_server.tls_peer_state in [TLSState.TO_TLS, TLSState.TLS]:
            # replace http with https in listen-peer-urls, initial-cluster and initial-advertise-peer-urls
            config_properties["listen-peer-urls"] = self.state.unit_server.peer_url.replace(
                "http://", "https://"
            )
            # in the `rebuild_cluster` workflow `initial-cluster` is not set
            if config_properties.get("initial-cluster"):
                config_properties["initial-cluster"] = config_properties[
                    "initial-cluster"
                ].replace(
                    self.state.unit_server.peer_url,
                    self.state.unit_server.peer_url.replace("http://", "https://"),
                )
            config_properties["initial-advertise-peer-urls"] = config_properties[
                "initial-advertise-peer-urls"
            ].replace("http://", "https://")

            # set the peer-transport-security
            config_properties["peer-transport-security"] = {
                "cert-file": self.workload.paths.tls.peer_cert.as_posix(),
                "key-file": self.workload.paths.tls.peer_key.as_posix(),
                "client-cert-auth": True,
                "trusted-ca-file": self.workload.paths.tls.peer_ca.as_posix(),
            }

        if self.state.unit_server.tls_client_state == TLSState.TO_NO_TLS:
            config_properties["listen-client-urls"] = self.state.unit_server.client_url.replace(
                "https://", "http://"
            )
            config_properties["advertise-client-urls"] = self.state.unit_server.client_url.replace(
                "https://", "http://"
            )

        if self.state.unit_server.tls_peer_state == TLSState.TO_NO_TLS:
            config_properties["listen-peer-urls"] = self.state.unit_server.peer_url.replace(
                "https://", "http://"
            )
            # in the `rebuild_cluster` workflow `initial-cluster` is not set
            if config_properties.get("initial-cluster"):
                config_properties["initial-cluster"] = config_properties[
                    "initial-cluster"
                ].replace("https://", "http://")
            config_properties["initial-advertise-peer-urls"] = config_properties[
                "initial-advertise-peer-urls"
            ].replace("https://", "http://")

        return yaml.safe_dump(config_properties)

    def set_config_properties(self) -> None:
        """Write the config properties to the config file."""
        logger.debug("Writing configuration")
        self.workload.write_file(
            content=self.config_properties,
            path=self.config_file,
        )

    def are_tuning_parameters_valid(self) -> bool:
        """Validate configuration values for tuning parameters.

        Returns:
            bool: True if tuning config values are valid, False if invalid.
        """
        if not self._are_timeout_parameters_valid():
            return False

        if not self._is_backend_quota_valid():
            return False

        if not self._are_compaction_parameters_valid():
            return False

        return True

    def _are_timeout_parameters_valid(self) -> bool:
        """Validate configuration for heartbeat-interval and election timeout."""
        if heartbeat_interval := self.config.get(TuningOptions.HEARTBEAT_INTERVAL_CONFIG.value):
            if heartbeat_interval < 10 or heartbeat_interval > 5000:
                logger.error(
                    "Heartbeat interval %s is invalid. It must be between 10ms and 5000ms.",
                    heartbeat_interval,
                )
                return False

        if election_timeout := self.config.get(TuningOptions.ELECTION_TIMEOUT_CONFIG.value):
            if election_timeout < heartbeat_interval * 10 or election_timeout > 50000:
                logger.error(
                    "Election timeout %s is invalid. It must be at least 10x the heartbeat interval (%s) and no more than 50000ms.",
                    election_timeout,
                    heartbeat_interval,
                )
                return False

        return True

    def _is_backend_quota_valid(self) -> bool:
        """Validate configuration for backend quota."""
        # avoid recalculating quota_backend_bytes multiple times
        quota_backend_bytes = self.quota_backend_bytes
        if quota_backend_bytes is None:
            logger.error(
                "Quota backend bytes value '%s' is invalid. It must be an integer or 'auto'.",
                self.config.get(TuningOptions.QUOTA_BACKEND_BYTES_CONFIG.value),
            )
            return False

        if quota_backend_bytes < MIN_QUOTA_BACKEND_BYTES:
            logger.error(
                "Quota backend bytes too low: %s. Minimum is %s.",
                quota_backend_bytes,
                MIN_QUOTA_BACKEND_BYTES,
            )
            return False

        if quota_backend_bytes < (db_file_size := self.workload.get_db_file_size()):
            logger.error(
                "Quota backend bytes %s is less than current DB file size %s",
                quota_backend_bytes,
                db_file_size,
            )
            return False

        if self.workload.is_lxd_cloud():
            logger.warning(
                "The deployment's quota-backend-bytes value is %s - consider applying constraints and/or setting the right lxd storage driver",
                quota_backend_bytes,
            )

        return True

    def _are_compaction_parameters_valid(self) -> bool:
        """Validate configuration for auto-compaction."""
        compaction_mode = str(self.config.get(TuningOptions.AUTO_COMPACTION_MODE.value))
        compaction_retention = str(self.config.get(TuningOptions.AUTO_COMPACTION_RETENTION.value))

        if compaction_mode not in ["periodic", "revision"]:
            logger.error("Auto-compaction-mode is invalid, must be 'periodic' or 'revision'")
            return False

        # accept numerical values or numerical values ending with "h" or "m"
        pattern = re.compile(r"^\d+[hm]?$")
        if not pattern.findall(compaction_retention):
            logger.error(
                "Auto-compaction-retention invalid, must be numerical value or 'h' for hours or 'm' for minutes"
            )
            return False

        if compaction_mode == "periodic" and not compaction_retention.endswith(("h", "m")):
            logger.error(
                "Auto-compaction-retention invalid, must be 'h' for hours or 'm' for minutes"
            )
            return False

        if compaction_mode == "revision" and compaction_retention.endswith(("h", "m")):
            logger.error("Auto-compaction-retention invalid, must be numerical in revision mode")
            return False

        if compaction_retention == "0" or compaction_retention.strip("h").strip("m") == "0":
            logger.error("Auto-compaction-retention is invalid, must not be 0")
            return False

        return True

    def requires_restart(self) -> bool:
        """Check current configuration and determine if restart is required."""
        try:
            current_config_values = self.workload.load_yaml_file(self.workload.paths.config_file)
        except yaml.YAMLError as e:
            logger.error(f"Error loading current config: {e}")
            return False

        for option in TuningOptions:
            if self._get_tuning_config_value(option) != current_config_values[option.value]:
                logger.info(f"Config change to {option.value} requires restart")
                return True

        return False

    def get_statuses(self, scope: Scope, recompute: bool = False) -> list[StatusObject]:
        """Compute the Cluster manager's statuses."""
        status_list: list[StatusObject] = []

        if not self.are_tuning_parameters_valid():
            status_list.append(ConfigStatuses.TUNING_CONFIG_INVALID.value)

        return status_list if status_list else [CharmStatuses.ACTIVE_IDLE.value]

    def _get_cluster_endpoints(self) -> str:
        """Concatenate peer-urls of all cluster members.

        Returns:
            str: Member name and peer url for all cluster members in required syntax, e.g.:
            etcd1=http://10.54.237.109:2380,etcd2=http://10.54.237.57:2380
        """
        cluster_endpoints = ",".join(
            f"{server.member_name}={server.peer_url}" for server in self.state.servers
        )

        return cluster_endpoints

    def _get_tuning_config_value(self, option: TuningOptions) -> int:
        """Get the value of a tuning configuration option.

        Args:
            option (TuningOptions): The tuning option to get the value for.

        Returns:
            int: The value of the tuning option.
        """
        if option == TuningOptions.QUOTA_BACKEND_BYTES_CONFIG:
            return self.quota_backend_bytes
        return self.config.get(option.value)

    @property
    def quota_backend_bytes(self) -> int | None:
        """Get the quota-backend-bytes configuration value.

        Returns:
            int: The quota-backend-bytes value.
        """
        config_value = self.config.get(TuningOptions.QUOTA_BACKEND_BYTES_CONFIG.value)
        if config_value == "auto":
            memory_max = int(self.workload.memory_size() * 0.9)
            if self.workload.data_storage_attached():
                storage_max = int(self.workload.data_storage_size() * 0.9)
            else:
                storage_max = MAX_QUOTA_BACKEND_BYTES
            value = min(
                MAX_QUOTA_BACKEND_BYTES,
                memory_max,
                storage_max,
            )
            if value == memory_max:
                logger.warning(f"Quota backend bytes reduced to {value} to fit available memory.")
            if value == storage_max:
                logger.warning(f"Quota backend bytes reduced to {value} to fit available storage.")
            return value

        try:
            return int(str(config_value))
        except (TypeError, ValueError):
            return None
